src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -4253,91 +4253,102 @@ protected virtual void Dispose(bool disposing)
                 return;
             }
 
-            // Close the handle in one of several ways depending on the timeout.
-            // Ignore ObjectDisposedException just in case the handle somehow gets disposed elsewhere.
-            try
+            if (!disposing)
             {
-                int timeout = _closeTimeout;
-                if (timeout == 0 || !disposing)
-                {
-                    // Abortive.
-                    if (NetEventSource.Log.IsEnabled()) NetEventSource.Info(this, "Calling _handle.CloseAsIs()");
-                    _handle?.CloseAsIs(abortive: true);
-                }
-                else
+                // When we are running on the finalizer thread, we don't call CloseAsIs
+                // because it may lead to blocking the finalizer thread when trying
+                // to abort on-going operations. We directly dispose the SafeHandle.
+                if (NetEventSource.Log.IsEnabled()) NetEventSource.Info(this, "Calling _handle.Dispose()");
+                _handle.Dispose();
```

Comment:
```suggestion
                if (_handle != null)
                { 
                    if (NetEventSource.Log.IsEnabled()) NetEventSource.Info(this, "Calling _handle.Dispose()");
                    _handle.Dispose();
                }    
```

Reply:
I combined it with the earlier check so also non-owning handles end up calling `DisposeCachedTaskSocketAsyncEventArgs`.